pkg/pma/pma.c

```diff
@@ -0,0 +1,901 @@
+}
+
+/// Sync changes made to either the heap or stack since the last call to
+/// pma_sync().
+///
+/// @param[in] path        Path to backing file. If NULL, this function is a
+///                        no-op.
+/// @param[in] base        Base address of the file-backed mapping.
+/// @param[in] grows_down  true if the mapping grows downward in memory.
+/// @param[in] pma         PMA this mapping belongs to.
+/// @param[in] len         Number of bytes at the beginning of the mapping to be
+///                        synced to the backing file. Must be a multiple of
+///                        kPageSz.
+/// @param[in] fd          Open file descriptor of the backing file.
+///
+/// @return 0   Synced changes successfully.
+/// @return -1  Failed to sync changes.
+static int
+_sync_file(const char *path,
+           void       *base,
+           bool        grows_down,
+           pma_t      *pma,
+           size_t      len,
+           int         fd);
+
+/// Get the total length in bytes of a PMA.
+///
+/// @param[in] pma
+static inline size_t
+_total_len(const pma_t *pma)
+{
+    assert(pma);
+    return (size_t)(pma->stack_start - pma->heap_start);
+}
+
+static int
+_handle_page_fault(void *fault_addr, void *user_arg)
+{
+    fault_addr = (void *)round_down((uintptr_t)fault_addr, kPageSz);
+    pma_t *pma = user_arg;
+    switch (_page_status(fault_addr, pma)) {
+        case PS_UNMAPPED:
+            if (MAP_FAILED
+                == mmap(fault_addr,
+                        kPageSz,
+                        PROT_READ | PROT_WRITE,
+                        MAP_ANON | MAP_FIXED | MAP_PRIVATE,
+                        -1,
+                        0))
+            {
+                fprintf(stderr,
+                        "pma: failed to create %zu-byte anonymous mapping at "
+                        "%p: %s\r\n",
+                        kPageSz,
+                        fault_addr,
+                        strerror(errno));
+                return 0;
+            }
+            _set_page_status(fault_addr, PS_MAPPED_DIRTY, pma);
+            break;
+        case PS_MAPPED_CLEAN:
+            if (mprotect(fault_addr, kPageSz, PROT_READ | PROT_WRITE) == -1) {
+                fprintf(stderr,
+                        "pma: failed to remove write protections from %zu-byte "
+                        "page at %p: %s\r\n",
+                        kPageSz,
+                        fault_addr,
+                        strerror(errno));
+                return 0;
+            }
+            _set_page_status(fault_addr, PS_MAPPED_DIRTY, pma);
+            break;
+        case PS_MAPPED_DIRTY:
+            fprintf(stderr,
+                    "pma: unexpectedly received a page fault on a dirty page "
+                    "at %p\r\n",
+                    fault_addr);
+            exit(EFAULT);
+        case PS_MAPPED_INACCESSIBLE:
+            // Center guard page.
+            if (pma_adjust(pma) == -1) {
+                return 0;
+            }
```

Comment:
If I'm reading this correctly, hitting the guard page will result in a segfault that crashes the process. Is that correct?